File: src/main/java/ru/ioleksiv/telegram/bot/core/controller/annotations/parser/unpacker/CallbackQueryUnpackerParser.java
package ru.ioleksiv.telegram.bot.core.controller.annotations.parser.unpacker;

import org.springframework.stereotype.Component;
import ru.ioleksiv.telegram.bot.api.annotations.receiver.CallbackQueryReceiver;
import ru.ioleksiv.telegram.bot.api.model.objects.std.CallbackQuery;
import ru.ioleksiv.telegram.bot.core.controller.handler.unpack.UpdateUnpacker;

import java.util.Optional;

@Component
public class CallbackQueryUnpackerParser extends UnpackerParser<CallbackQueryReceiver, CallbackQuery> {

    @Override
    public Class<CallbackQueryReceiver> getAnnotationClass() {
        return CallbackQueryReceiver.class;
    }

    @Override
    public UpdateUnpacker<CallbackQuery> getUpdateUnpacker() {
        return new UpdateUnpacker<>(in -> Optional.ofNullable(in.getCallbackQuery()), CallbackQuery.class);
    }

}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/core/controller/annotations/parser/unpacker/CallbackQueryUnpackerParser.kt
package io.github.oleksivio.tl.kbot.core.core.controller.annotations.parser.unpacker

import io.github.oleksivio.tl.kbot.core.annotations.receiver.CallbackQueryReceiver
import io.github.oleksivio.tl.kbot.core.core.controller.handler.unpack.UpdateUnpacker
import io.github.oleksivio.tl.kbot.server.api.objects.std.CallbackQuery

class CallbackQueryUnpackerParser : UnpackerParser<CallbackQueryReceiver, CallbackQuery>() {
    override val updateUnpacker: UpdateUnpacker<CallbackQuery>
        get() = UpdateUnpacker(
            { it.callbackQuery },
            CallbackQuery::class
        )

    override val parserAnnotationClass = CallbackQueryReceiver::class
}
